File: perplexor/remultiplexor/remultiplexor-paired.pl
#!/usr/bin/perl -w

# Writen by I. Lagkouvards (ilias.lagkouvardos@tum.de)

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

use strict;
use warnings;


#Declare variables
my $barcodesfilename = "barcodes.txt";
my $demultiplexed_pairs = "multiplexing_table.tab";
my $I1_filename= "I1.fastq";
my $R1_filename= "R1.fastq";
my $R2_filename= "R2.fastq";
my $mappingfile= "mapping_file.tab";
my $files_folder= "samples-sequences";

#Open file with the illumina barcodes
open (my $barcodes_fh,"<",$barcodesfilename) || die "Cannot open \"$barcodesfilename\" to read from: $!";

#parse barcodes to barcodes array
my @barcodes;
while(my $barcode_line = <$barcodes_fh>)
{
   chomp $barcode_line;
   push (@barcodes,$barcode_line);
}

#open the files that we will write the merged output
open(my $I1_fh,">", $I1_filename) || die "Cannot open \"$I1_filename\" to write to: $!";
open(my $R1_fh,">", $R1_filename) || die "Cannot open \"$R1_filename\" to write to: $!";
open(my $R2_fh,">", $R2_filename) || die "Cannot open \"$R2_filename\" to write to: $!";

# expected format for the pairs filename
# samplename\tforward_read_filename\treverse_read_filename\n
# sample1  sample1_R1.fastq  sample1_R2.fastq
open (my $pairs_fh,"<",$demultiplexed_pairs) || die "Cannot open \"$demultiplexed_pairs\" to read from: $!";
# Open the file to write the mappings
open (my $mappings_fh,">",$mappingfile) || die "Cannot open \"$mappingfile\" to write to: $!";
print $mappings_fh "#SampleID\tIndex\n";

my $pairnumber=0;
while (my $pair_line= <$pairs_fh>)
{
  chomp $pair_line;
  if ($pair_line=~/^#/) { next;}  
  my ($sample_name,$forward_filename,$reverse_filename) = split (/\t/,$pair_line);
  open (my $forward_fh,"<","./$files_folder/$forward_filename") || die "Cannot open \"$forward_filename\" to read from: $!";
  open (my $reverse_fh,"<","./$files_folder/$reverse_filename") || die "Cannot open \"$reverse_filename\" to read from: $!";
  do
  {
    #read the paired reads from the two read files of one sample
    my @f_read_entry=getfastq($forward_fh);
    my @r_read_entry=getfastq($reverse_fh);

    #create a pseudo index entry for each sequence pair using the same barcode for the whole sample
    my @index_entry=();
    push (@index_entry,"$f_read_entry[0]");
    push (@index_entry,"$barcodes[$pairnumber]");
    push (@index_entry,"+");
    push (@index_entry,"EEEEEEEEEEEE");
    
    printfastq($I1_fh,@index_entry);    
    printfastq($R1_fh,@f_read_entry);
	printfastq($R2_fh,@r_read_entry);
  }
  until eof($forward_fh);
  close ($forward_fh);
  close ($reverse_fh);
  
  print $mappings_fh "$sample_name\t$barcodes[$pairnumber]\n";
  $pairnumber++;    
}

################## SUBS ######################
#get an entry from a fastq file
sub getfastq
{
  my($filehandler)= @_;
  my @entry=();
  for (my $x=0;$x<4;$x++)
  {
    my $line = <$filehandler>;
    chomp $line;
    push (@entry,$line);
  }
  return (@entry);
}

#print a fastq entry
sub printfastq
{
  my($filehandler,@fastq_entry)= @_;
  foreach my $line (@fastq_entry)
  {
    print $filehandler "$line\n";
  }  
}

File: perplexor/remultiplexor/remultiplexor-paired.py
# Writen by I. Lagkouvards (ilias.lagkouvardos@tum.de)

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

# import libraries
import re
import os
import sys
import common

# declare variables
I1_filename= "I1.fastq"
R1_filename= "R1.fastq"
R2_filename= "R2.fastq"

pairs_filename = "multiplexing_table.tab"
barcodes_filename = "barcodes.txt"
mappings_filename= "mapping_file.tab"

files_folder= "samples-sequences"

# read barcodes from file
barcodes = common.read_barcodes(barcodes_filename)

# read pairs from file
pairs_fh = common.open_file_else_exit(pairs_filename, 'r')
pair_lines = pairs_fh.readlines()

# Open file to write  mappings
mappings_fh = common.open_file_else_exit(mappings_filename, 'w')
mappings_fh.write("#SampleID\tIndex\n")

# create new Index, R1, R2 files 
I1_fh = common.open_file_else_exit(I1_filename, 'w')
R1_fh = common.open_file_else_exit(R1_filename, 'w')
R2_fh = common.open_file_else_exit(R2_filename, 'w')

pair_counter = 0
for pair_line in pair_lines:

  pair_line = pair_line.rstrip('\n')
  if pair_line.startswith('#'):
    continue
  
  if not pair_line:
    continue
  
  splits = pair_line.split("\t")
  if len(splits) < 3:
    continue

  sample_name = splits[0]
  forward_filename = splits[1]
  reverse_filename = splits[2]

  forward_full_path_name = './{}/{}'.format(files_folder, forward_filename)
  reverse_full_path_name = './{}/{}'.format(files_folder, reverse_filename)

  forward_fh = common.open_file_else_exit(forward_full_path_name, 'r')
  reverse_fh = common.open_file_else_exit(reverse_full_path_name, 'r')

  forward_entry = []
  reverse_entry = []
  line_counter = 0

  for forward_line in forward_fh:

    forward_line = forward_line.rstrip('\n')
    reverse_line = reverse_fh.readline().rstrip('\n')

    line_counter = line_counter + 1
      
    forward_entry.append(forward_line)
    reverse_entry.append(reverse_line)

    if line_counter == 4:
      index_entry = []
      index_entry.append(forward_entry[0])
      index_entry.append(barcodes[pair_counter])
      index_entry.append("+")
      index_entry.append("EEEEEEEEEEEE")
        
      common.write_fastq_entry_to_file(I1_fh, index_entry)
      common.write_fastq_entry_to_file(R1_fh, forward_entry)
      common.write_fastq_entry_to_file(R2_fh, reverse_entry)
      
      forward_entry = []
      reverse_entry = []
      line_counter = 0

  mappings_fh.write('{}\t{}\n'.format(sample_name, barcodes[pair_counter]))
  
  forward_fh.close()
  reverse_fh.close()

  pair_counter = pair_counter + 1
  